### ISSUE TITLE
Can't force push a ref to a remote if local repo doesn't have the commit referred to by the current remote ref

### ISSUE THREAD

**Issue by karlhendrikse** [git-lfs/git-lfs#4140]
**Describe the bug**
This is a corner case, but one I've run into frequently over the last few days. When trying to force push, if the local repository doesn't contain the commit that the remote has for the ref being pushed, LFS causes an error and the push fails.

It's fairly easy to run into this when migrating a repository to LFS. For example: push, migrate, GC (to see the effect the migration has had on the repo size), force push. (Or more commonly for me, migrate, push, clone to test the performance, migrate a few more file extensions, force push, clone again...). Below are synthetic steps that achieve a similar thing.

**To reproduce**
```
# Create a "remote" repository.
mkdir remote
cd remote
git init --bare .
cd ..

# Create a "local" repository.
mkdir local
cd local
git clone file:///c/Data/Code/test/remote .
git lfs install --local # <----------------- The error only happens if this line is present.

# Add a commit "A".
#
# A <- master
touch a
git add a
git commit -m "add a"

# Add another commit "B".
#
# B <- master
# |
# A
touch b
git add b
git commit -m "add b"

# Push.
#
# B <- master, origin/master
# |
# A
git push origin master

# Undo the last commit.
#
# B <- origin/master
# |
# A <- master
git reset --hard HEAD~

# Add a different commit "C" instead.
#
# B <- origin/master
# |
# | C <- master
# |/
# A
touch c
git add c
git commit -m "add c"

# Make sure that we no longer have commit B at all, even though that's what the remote
# repository's master branch is pointing to.
#
# C <- master
# |
# A
git remote remove origin
git reflog expire --expire=all --all
git gc --prune=now
git remote add origin file:///c/Data/Code/test/remote

# Try to (force) push again. This produces an error.
git push -f origin master
```

**Observed behavior**
```
ref master:: Error in git rev-list --stdin --objects --not --remotes=origin --: exit status 128 fatal: bad object bd2e3a29aca555968b908157f3bcd74100593d28

error: failed to push some refs to 'file:///c/Data/Code/test/remote'
```
(That hash in the error message is commit B)

**Expected behavior**
The force push should succeed.

**System environment**
Git Bash, Windows 10.

**Output of `git lfs env`**
```
git-lfs/2.10.0 (GitHub; windows amd64; go 1.12.7; git a526ba6b)
git version 2.26.2.windows.1

LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs\objects
LocalReferenceDirs=
TempDir=lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Workaround**
This problem could be addressed by not doing the garbage collection, but A. if garbage collection has already been done it's too late, B. sometimes it's done by a tool and C. it should be valid to garbage collect whenever.

In any case, the solution I've used is to pull the ref in question from the remote, then run the push again. That pull can be quite sizeable (e.g. the entire pre-migration history of master that was just thrown away) so the workaround isn't ideal.

Another workaround is to delete the ref from the remote before doing the push. This can be a pain if you're trying to push many refs at a time.

**Comment by bk2204** [comment]
Hey,,

I believe we've fixed this in #4102, which should be in 2.11.0.  Could you verify that things work for you with that version?

**Comment by karlhendrikse** [comment]
Confirmed that this is fixed in 2.11.0.
